# DS-SF-36 | 01 | What is Data Science | Assignment | Starter Code

## Python

> ## "Data scientists' most basic, universal skill is the ability to write code."
>
> Harvard Business Review | “Data Scientists: The Sexiest Job of the 21st Century” (2012)
>
> (https://hbr.org/2012/10/data-scientist-the-sexiest-job-of-the-21st-century)

Today, to give you some more practice beyond the course pre-requisites, we are going to implement a few functions in Python.  This assignment covers to some extend the following topics:

- Functions (defining and using your own functions but also calling functions from packages)
- Loops
- Arithmetic operations
- Conditional statements

**Don't worry if you get stuck.  Ask around, review the answer key, and ask around more.  As this course progresses, your programming proficiency will increase.**

> ### Question 1.  Multiples of 3 and 5
>
> If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9.  The sum of these multiples is 23.
>
> Find the sum of all the multiples of 3 or 5 below 1,000.
>
> (Source: [Project Euler | Problem 1](https://projecteuler.net/problem=1))

In [6]:
# Create function called multi_3_5(n) to calculate a sum...
# where (i) n is below 10 and (ii) n is below 1,000
def multi_3_5(n):
    sum = 0
    # loop i where range(1, n); excludes 'n' so will iterate everything below 'n'
    for i in range(1, n):
        if (i % 3 == 0) or (i % 5 == 0):
            sum += i
    return sum

multi_3_5(10)
multi_3_5(1000)

233168

Answer: 233168

> ### Question 2.  Estimating square roots
>
> Given a real number $m$, let's define the series $u$ as follow:
> - $u_0 = 1$
> - $u_{n+1} = \frac{u_n ^ 2 + m}{2u_n}$
>
>
> Implement the calculations of the series $u$ above to estimate square roots.  Verify that $\sqrt{144} = 12$ and use your function to calculate $\sqrt{1024}$.

In [19]:
# Create function called sqr(m) to calculate square root
def sqr(m):
    # Creating u_n variable and setting it equal to 1 (adding '.' to indicate number?) 
    u_n = 1.
    
    while True:
        u_n_plus_1 = (u_n ** 2 + m) / (2 * u_n)
        # absolute value and less than 10 to the minus 6 then break loop (hmm, not 100% sure what's going on here)
        # assigning decimal places?
        if abs(u_n_plus_1 / u_n - 1) < 10 ** -6:
            break
        # edited u_n with u_n_plus_1; set equal
        u_n = u_n_plus_1
    return u_n

In [20]:
sqr(144)

12.000000012408687

In [16]:
sqr(1024)

32.0000071648159

Answer: TODO

> ### Question 3.  Prime Numbers
>
> A prime (number) is a natural number greater than 1 that has no positive divisors other than 1 and itself.  ([Wikipedia](https://en.wikipedia.org/wiki/Prime_number))
>
> Calculate all primes below 1,000.  What's their sum?

In [37]:
import math

In [38]:
math.sqrt(1024)

32.0

In [44]:
# Step 1: define a prime number
# Loop i where n between 2 and hmm... 'no positive divisors other than 1 and itself'
# and loop i where if n is divided by itself and it's zero then it's excluded (False)
def prime(n):
    for i in range(2, int(sqr(n)) + 1):
        if n % i == 0:
            return False
    return True

# Step 2: Calculate primes < 1000 
# Create an empty list
# Loop i where n is between 2 and prime plus 1 since it will include the highest prime and exclude n + 1
# Redefine the prime argument i using append operation
def prime_sum(n):
    l = []
    for i in range(2, n + 1):
        if prime(i)
            l.append(i)
    return l

# Step 3: calculating the sum of primes below 1,000 by calling the 'l' list
l = prime_sum(100)

In [45]:
sum(l)

1060

Answer: TODO

> ### Question 4.  Largest prime factor
>
> The prime factors of 13195 are 5, 7, 13 and 29.
>
> What is the largest prime factor of the number 600851475143?
>
> (Source: [Project Euler | Problem 3](https://projecteuler.net/problem=3))

In [47]:
# Define largest prime factor
# Hmm, totally lost...

def largest_prime_factor(n):
    primes = set()

    i = 2
    while n > 1:
        while n % i == 0:
            primes.add(i)
            n /= i
        i += 1

    return max(primes)

In [48]:
largest_prime_factor(13195)

29

In [50]:
largest_prime_factor(600851475143)

6857

Answer: TODO

> ### Question 5.  Mean
>
> Write a function to calculate the mean (average) of a list.
>
> What's the mean of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [57]:
# Step 1: define mean (average) of a list, 'sum of numbers' divided by 'number of numbers'
# Let x equal to the number of numbers
# Denominator can't be zero
# Reset "sum" variable to zero.
def avg(n):
    x = len(n)
    if x == 0:
        return False
    
    sum = 0.
    
    # Loop i where n is equal to the sum of n's in the list
    # Assign sum to new sum
    for i in n:
        sum += i
    
    # Call sum divided by number of numbers
    return sum / x

In [58]:
l = [10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5]

avg(l)

9.0

In [59]:
import numpy as np

np.mean(l)

9.0

Answer: TODO

> ### Question 6.  Sample standard deviation
>
> Write a function to calculate the standard deviation of a sample.
>
> Given the sample $x_1, x_2, ..., x_N$, its standard deviation is defined as $s = \sqrt{\frac{1}{N - 1} \sum_{i = 1}^{N} (x_i  - \bar{x})^2}$, with $\bar{x}$ as the sample mean.
>
> What's the standard deviation of the following sample: 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?
>
> ([Wikipedia](https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation))

In [68]:
# Create a function to calculate the standard deviation of a sample
# Find the standard deviation of the following sample: 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5
def stdev(n):
    # Again, the denominator cannot equal zero
    x = len(n)
    if n <= 1:
        return None

    sum = 0.

    # Call previous function avg(n) 
    n_bar = avg(n)

    # Loop 
    for i in n:
        sum += (i - n_bar) ** 2

    return sqr(sum / (x - 1))

In [69]:
l = [10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5]

stdev(l)

3.3166248052315686

In [70]:
np.std(l, ddof = 1)

3.3166247903553998

Answer: TODO

> ### Question 7.  Median
>
> Write a function to calculate the median ("middle value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Median))
>
> What's the median of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [72]:
# Sort values in a list and then find the median
def median(n):
    # Make sure the list has 1 or more values and if zero then return 'None'
    x = len(n)
    if x == 0:
        return None
    
    # Sort the list
    sorted_n = sorted(n)
    
    # Divide the length of list n into 2 by calling x
    i = x / 2

    # Lost from here... hmmm
    if (x % 2 == 1):
        return sorted_n[i]
    else:
        return avg(sorted_n[i - 1: i + 1])

In [73]:
np.median(l)

9.0

Answer: TODO

> ### Question 8.  Mode
>
> Write a function to calculate the mode ("most frequent value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Mode_(statistics)))
>
> What's the mode of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7 and 5?  How about the mode of 8, 8, 8, 8, 8, 8, 19, 8, 8 and 8?

In [ ]:
# Create function to calculate mode (most frequent value)
# Calculate the mode of a list of numbers
def mode(n):
    # Create dictionary counts
    # Loop i for n where it counts in increments of 1
    # or the count is 1
    counts = {}
    for i in n:
        if i in counts:
            counts[i] += 1
        else:
            counts[i] = 1
    
    # Hmm, why?
    mode_count = 2
    # Creating empty list for mode_n and will append mode_n
    mode_n = []

    # Loop count for counts? 
    for i, count in counts.iteritems(): # counts.items() in Python 3
        if mode_count < count:
            mode_count = count
            mode_n = [i]
        elif count == mode_count:
            mode_n.append(i)

    return mode_n

Answer: TODO